sno/working_copy/postgis.py

```diff
@@ -116,83 +119,113 @@ def session(self, bulk=0):
+            scale = extra_type_info.get("scale", None)
+            if precision is not None and scale is not None:
+                return f"numeric({precision},{scale})"
+            elif precision is not None:
+                return f"numeric({precision})"
+            else:
+                return "numeric"
+
+        return pg_type
 
     def _get_columns(self, dataset):
         pk_field = None
-        cols = {}
-        sqlite_cols = dataset.get_meta_item("sqlite_table_info")
-        for col in sqlite_cols:
-            gpkg_type = col["type"]
-            if gpkg_type in (
-                "POINT",
-                "LINESTRING",
-                "POLYGON",
-                "MULTIPOINT",
-                "MULTILINESTRING",
-                "MULTIPOLYGON",
-                "GEOMETRYCOLLECTION",
-            ):
-                srid = dataset.get_meta_item("gpkg_geometry_columns")["srs_id"]
-                pg_type = f"GEOMETRY({gpkg_type}, {srid})"
-            elif gpkg_type in ("GEOMETRY"):
-                srid = dataset.get_meta_item("gpkg_geometry_columns")["srs_id"]
-                pg_type = f"GEOMETRY(GEOMETRY, {srid})"
-            elif gpkg_type in ("MEDIUMINT", "INT8"):
-                pg_type = "integer"
-            elif gpkg_type in ("TINYINT", "INT2"):
-                pg_type = "smallint"
-            elif gpkg_type in ("UNSIGNED BIG INT"):
-                pg_type = "bigint"
-            elif gpkg_type in ("FLOAT"):
-                pg_type = "real"
-            elif gpkg_type in ("DOUBLE"):
-                pg_type = "double precision"
-            elif gpkg_type in ("DATETIME"):
-                pg_type = "timestamptz"
-            elif re.match(
-                r"((N?((ATIVE )|(VAR(YING )?))?CHAR(ACTER)?)|TEXT|CLOB)(\(\d+\))?$",
-                gpkg_type,
-            ):
-                m = re.split(r"[\(\)]", gpkg_type)
-                pg_type = f"VARCHAR({m[1]})" if len(m) == 3 else "TEXT"
-            else:
-                # FIXME
-                pg_type = gpkg_type
-
+        pg_cols = {}
+        for col in dataset.schema:
+            pg_type = self._v2_type_to_pg_type(dataset, col)
             col_spec = [
-                Identifier(col["name"]),
+                Identifier(col.name),
                 SQL(pg_type),
             ]
-            if col["pk"]:
-                col_spec.append(SQL("PRIMARY KEY"))
-                pk_field = col["name"]
-            if col["notnull"]:
-                col_spec.append(SQL("NOT NULL"))
-            cols[col["name"]] = SQL(" ").join(col_spec)
+            if col.pk_index is not None:
+                col_spec.append(SQL("PRIMARY KEY NOT NULL"))
+                pk_field = col.name
+            pg_cols[col.name] = SQL(" ").join(col_spec)
 
-        self.L.debug("Sqlite>Postgres column mapping: %s -> %s", sqlite_cols, cols)
-        return cols, pk_field
+        self.L.debug(
+            "Sqlite>Postgres column mapping: %s -> %s", dataset.schema, pg_cols
```

Comment:
we're not really mapping sqlite here anymore :)

Reply:
Done